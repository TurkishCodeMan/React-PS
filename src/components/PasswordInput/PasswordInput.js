import PropTypes from "prop-types";
import EyeIcon from "../EyeIcon";
import TextInput from "../TextInput";
import ProgressBar from "../ProgressBar";
import { useState } from "react";

function PasswordInput({
    htmlId,
    name,
    label,
    onChange,
    onBlur,
    placeholder,
    value,
    error,
    showVisibilityToggle,
    quality,
    minLength,
    maxLength,
    ...props
}) {
    const [showPassword, setShowPassword] = useState(false);

    function toggleShowPassword() {
        return setShowPassword((prevState) => {
            return !prevState
        })
    }

    return (
        <TextInput
            htmlId={htmlId}
            name={name}
            label={label}
            placeholder={placeholder}
            type={showPassword ? 'text' : 'password'}
            onChange={onChange}
            onBlur={onBlur}
            value={value}
            maxLength={maxLength}
            error={error}
            required
            {...props}
        >
            {showVisibilityToggle &&
                <a
                    aria-label="toggle"
                    onClick={toggleShowPassword}
                    style={{ marginLeft: 5 }}
                >
                    <EyeIcon />
                </a>
            }

            {
                value.length > 0 && quality && <ProgressBar percent={quality} width={130} />
            }
        </TextInput>
    )
}

PasswordInput.propTypes = {
    /** Unique HTML ID. Used for tying label to HTML input. Handy hook for automated testing. */
    htmlId: PropTypes.string.isRequired,

    /** Input name. Recommend setting this to match object's property so a single change handler can be used. */
    name: PropTypes.string.isRequired,

    /** Input label */
    label: PropTypes.string.isRequired,

    /** Function to call onChange */
    onChange: PropTypes.func.isRequired,

    /** Placeholder to display when empty */
    placeholder: PropTypes.string,

    /** Value */
    value: PropTypes.any,

    /** String to display when error occurs */
    error: PropTypes.string,

    /** Dislpay progressbar */
    quality: PropTypes.number,

    showVisibilityToggle: PropTypes.bool,

    maxLength: PropTypes.number,
    minLength: PropTypes.number
};

PasswordInput.defaultProps = {
    maxLength: 50,
    showVisibilityToggle: false,
    label: 'Password',
    name: "default"
}

export default PasswordInput;