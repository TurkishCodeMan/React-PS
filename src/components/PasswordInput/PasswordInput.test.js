import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import PasswordInput from "./PasswordInput";
import userEvent from "@testing-library/user-event";

test("hides pass quality by default", () => {
    const tree = renderer.create(<PasswordInput
        htmlId="test"
        name="test"
        onChange={() => { }}
        value="Uisi44"

    />).toJSON();
    expect(tree).toMatchSnapshot();
});


test("toggle input type show/hide pass clicked", () => {
    render(<PasswordInput
        htmlId="test"
        name="test"
        value="aa"
        placeholder="pass"
        onChange={() => { }}
        showVisibilityToggle
    />)

    const toggleArea = screen.getByLabelText('toggle');
    const ps = screen.getByPlaceholderText('pass');
    expect(toggleArea).toBeInTheDocument();
    expect(ps.type).toMatch('password');
    userEvent.click(toggleArea);
    expect(ps.type).toMatch('text')
})