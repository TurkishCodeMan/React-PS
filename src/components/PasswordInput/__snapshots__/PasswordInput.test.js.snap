// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hides pass quality by default 1`] = `
<div
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <label
    htmlFor="test"
    style={
      Object {
        "display": "block",
      }
    }
  >
    Password
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
       *
    </span>
  </label>
  <input
    id="test"
    maxLength={50}
    name="test"
    onChange={[Function]}
    type="password"
    value="Uisi44"
  />
</div>
`;
